sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/FunMatcher.kt

```diff
@@ -107,8 +109,15 @@ class FunMatcher(
         when (functionDescriptor) {
             is ConstructorDescriptor -> false
             else -> {
-                functionDescriptor.overriddenDescriptors.any {
-                    it?.fqNameSafe?.asString()?.substringBeforeLast(".") == supertype
+                // Note: `overriddenTreeUniqueAsSequence()` will return a sequence of all overridden descriptors, even if thery are not
+                // explicitly declared in the super types. In other words, if you have a class hierarchy of 10 levels, with a declaration
+                // in the top-most class/interface and an overriding method in the bottom-most, it will return a sequence of 10 elements,
+                // even if there are no overriding methods in the classes in-between. This can be desired but in most cases is somewhat
+                // inefficient, as we mostly look for the top-most declaration here. The top-most declaration is the last element in the
+                // sequence, however, so we need to iterate the entire sequence to get there. There are optimization possibilities if this
+                // turns out to be a performance problem in the long run.
+                functionDescriptor.overriddenTreeUniqueAsSequence(true).any {
```

Comment:
As far as I understood this code is only triggered when we explicitly mention that we want to match the supertype.
I remember we discussed making it the default behaviour everywhere, so did you decide to not change it at the moment?

Reply:
Not for the moment, as it is not an easy fix. It's noted down for later :).